src/mr/master.go

```diff
@@ -1,29 +1,115 @@
 package mr
 
-import "log"
-import "net"
-import "os"
-import "net/rpc"
-import "net/http"
+import (
+	"log"
+	"net"
+	"net/http"
+	"net/rpc"
+	"os"
+	"sync"
+	"time"
+)
+
+type taskPool struct {
+	pending map[int]interface{}
+	issued  map[int]interface{}
+	mutex   sync.Mutex
+}
+
+func makeTaskPool() taskPool {
+	return taskPool{pending: make(map[int]interface{}), issued: make(map[int]interface{})}
+}
+
+func (tp *taskPool) insert(id int, task interface{}) {
+	tp.mutex.Lock()
+	defer tp.mutex.Unlock()
+	tp.pending[id] = task
+}
 
+func (tp *taskPool) issue() (bool, interface{}) {
+	tp.mutex.Lock()
+	defer tp.mutex.Unlock()
+	for id, task := range tp.pending {
+		delete(tp.pending, id)
+		tp.issued[id] = task
+		time.AfterFunc(10*time.Second, func() { tp.reset(id) })
```

Comment:
seems a bit odd about this pattern...
`tp.reset(id)` doesn't seem like at the same level of abstraction as the `issue()` func.

Reply:
oh hmm... on a second thought, this looks rather good!